Multitouch Support/Native/VoodooI2CNativeEngine.cpp

```diff
@@ -12,72 +12,92 @@
-}
-
-bool VoodooI2CNativeEngine::init(OSDictionary* properties) {
-    if (!super::init(properties))
-        return false;
-
-    simulator = OSTypeAlloc(VoodooI2CMT2SimulatorDevice);
-    actuator = OSTypeAlloc(VoodooI2CMT2ActuatorDevice);
-
-    if (!simulator || !actuator) {
-        OSSafeReleaseNULL(simulator);
-        OSSafeReleaseNULL(actuator);
-        return false;
+    if (!voodooInputInstance) {
+        return MultitouchReturnContinue;
     }
 
-    return true;
-}
-
-void VoodooI2CNativeEngine::free() {
-    OSSafeReleaseNULL(simulator);
-    OSSafeReleaseNULL(actuator);
-
-    super::free();
+    message.timestamp = timestamp;
+    message.contact_count = event.contact_count;
+    
+    for (int i = 0; i < event.contact_count; i++) {
+        VoodooI2CDigitiserTransducer* transducer = (VoodooI2CDigitiserTransducer*) event.transducers->getObject(i);
+        VoodooInputTransducer* inputTransducer = &message.transducers[i];
+        
+        if (!transducer) {
+            continue;
+        }
+        
+        inputTransducer->id = transducer->id;
+        inputTransducer->secondaryId = transducer->secondary_id;
+        
+        inputTransducer->type = VoodooInputTransducerType::FINGER;
+        
+        inputTransducer->isValid = transducer->is_valid;
+        inputTransducer->isTransducerActive = transducer->tip_switch.value();
+        inputTransducer->isPhysicalButtonDown = transducer->physical_button.value();
+        
+        inputTransducer->currentCoordinates.x = transducer->coordinates.x.value();
+        inputTransducer->previousCoordinates.x = transducer->coordinates.x.last.value;
+        
+        inputTransducer->currentCoordinates.y = transducer->coordinates.y.value();
+        inputTransducer->previousCoordinates.y = transducer->coordinates.y.last.value;
+        inputTransducer->supportsPressure = false;
+        inputTransducer->timestamp = timestamp;
+
+        // TODO: does VoodooI2C know width(s)? how does it measure pressure?
+        inputTransducer->currentCoordinates.width = transducer->tip_pressure.value() / 2;
+        inputTransducer->previousCoordinates.width = transducer->tip_pressure.last.value / 2;
+
+        inputTransducer->currentCoordinates.pressure = transducer->tip_pressure.value();
+        inputTransducer->previousCoordinates.pressure = transducer->tip_pressure.last.value;
+    }
+    
+    super::messageClient(kIOMessageVoodooInputMessage, voodooInputInstance, &message, sizeof(VoodooInputEvent));
+    
+    return MultitouchReturnBreak;
 }
 
 bool VoodooI2CNativeEngine::start(IOService* provider) {
     if (!super::start(provider))
         return false;
-
-    // Initialize simulator device
-    if (!simulator->init(NULL) || !simulator->attach(this))
-        goto exit;
-    else if (!simulator->start(this)) {
-        simulator->detach(this);
-        goto exit;
+    
+    parentProvider = NULL;
+    parentProvider = OSDynamicCast(VoodooI2CMultitouchInterface, provider);
```

Comment:
Doesn't `OSDynamicCast` return `nullptr` on failure?
```suggestion
    parentProvider = OSDynamicCast(VoodooI2CMultitouchInterface, provider);
```